File: 0399-evaluate-division/0399-evaluate-division.js
/**
 * @param {string[][]} equations
 * @param {number[]} values
 * @param {string[][]} queries
 * @return {number[]}
 */
var calcEquation = function(equations, values, queries) {
    const graph = new Map();
    
    // Build the graph
    for (let i = 0; i < equations.length; i++) {
        const [start, end] = equations[i];
        const value = values[i];
        
        if (!graph.has(start)) graph.set(start, {});
        if (!graph.has(end)) graph.set(end, {});
        
        graph.get(start)[end] = value;
        graph.get(end)[start] = 1 / value;
    }
    
    // Function for DFS
    const dfs = function(node, target, visited) {
        // If the node does not exist in the graph, return -1.0
        if (!graph.has(node)) return -1.0;
        // If the node is the target, return 1.0
        if (node === target) return 1.0;
        
        visited.add(node);
        
        const neighbors = graph.get(node);
        for (let nextNode in neighbors) {
            if (visited.has(nextNode)) continue;
            const product = dfs(nextNode, target, visited);
            
            if (product === -1.0) continue;
            
            return neighbors[nextNode] * product;
        }
        
        return -1.0;
    };
    
    // Run queries through DFS
    const results = [];
    for (let query of queries) {
        const [start, end] = query;
        results.push(dfs(start, end, new Set()));
    }
    
    return results;
};

File: 0399-evaluate-division/0399-evaluate-division.cpp
#include<vector>
#include<string>
#include<unordered_map>
using namespace std;

class Solution {
public:
    unordered_map<string, unordered_map<string, double>> g;
    
    double dfs(string start, string end, unordered_map<string, int>& visited) {
        if(g.find(start) == g.end() || visited[start]) 
            return -1.0;
        if(start == end) 
            return 1.0;
        
        visited[start] = 1;
        for(auto [node, weight]: g[start]) {
            double tmp = dfs(node, end, visited);
            if(tmp != -1.0) 
                return tmp * weight;
        }
        
        return -1.0;
    }

    vector<double> calcEquation(vector<vector<string>>& equations, vector<double>& values, vector<vector<string>>& queries) {
        for(int i = 0; i < equations.size(); i++) {
            string num = equations[i][0];
            string den = equations[i][1];
            double val = values[i];
            g[num][den] = val;
            g[den][num] = 1.0 / val;
        }

        vector<double> result;
        for(auto& q : queries) {
            string num = q[0];
            string den = q[1];
            unordered_map<string, int> visited;
            double tmp = dfs(num, den, visited);
            if(tmp != -1.0) 
                result.push_back(tmp);
            else 
                result.push_back(-1.0);
        }

        return result;
    }
};
